File: 2022/May/trie_structure.py
class Node:
    def __init__(self, key, data = None):
        self.key = key
        self.data = data
        self.children = {}


class Trie:
    def __init__(self):
        self.head = Node(None)
    
    def insert(self, string):
        curr_node = self.head

        # print('string: ', string)
        for char in string:
            # print('char: ', char)
            if char not in curr_node.children:
                curr_node.children[char] = Node(char)
                # print('new char: ', char)
            
            curr_node = curr_node.children[char]
            # print('currNode: ', curr_node)
            # print()
        # print('curr_node.key: ', curr_node.key)
        

        curr_node.data = string
        # print('curr_node.data: ', curr_node.data)

    def search(self, string):
        curr_node = self.head

        for char in string:
            if char in curr_node.children:
                curr_node = curr_node.children[char]
            
            else:
                return False
        
        if curr_node.data != None:

            return True
        else:
            return False

trie = Trie()
strings = ['apple', 'appolo', 'apptite', 'appstore', 'appdulla']

for string in strings:
    trie.insert(string)

print('apple: ', trie.search('apple'))

print('appgujeong: ', trie.search('appgujeong'))

File: 2022/May/trie_structure.js
// class Node {
//   constructor(key, data = null) {
//     this.key = key;
//     this.data = data;
//     this.children = {};
//   }
// }

// class Trie {
//   constructor() {
//     this.head = new Node(null);
//   }

//   insert(string) {
//     let currNode = this.head;

//     for (const char of string) {
//       if (!(char in currNode.children)) {
//         currNode.children[char] = new Node(char);
//       }

//       currNode = currNode.children[char];
//     }

//     currNode.data = string;
//   }

//   search(string) {
//     let currNode = this.head;

//     for (const char of string) {
//       if (char in currNode.children) {
//         currNode = currNode.children[char];
//       } else {
//         return false;
//       }
//     }

//     if (currNode.data === string) {
//       return true;
//     } else {
//       return false;
//     }
//   }
// }


// const trie = Trie();
// const stringArr = ['apple', 'appolo', 'apptite', 'appstore', 'appdulla'];

// for (const string in stringArr) {
//   trie.insert(string);
// }
  
// console.log('apple: ', trie.search('apple'));
// console.log('appgujeong: ', trie.search('apgujeong'));


const Node = function(key, data = null) {
  this.key = key;
  this.data = data;
  this.children = {};
}

const Trie = function() {
  this.head = new Node(null);

  this.insert = function(string) {
    let currNode = this.head;

    for (const char of string) {
      if (!(char in currNode.children)) {
        currNode.children[char] = new Node(char);
      }
    
      currNode = currNode.children[char];
    }
    
    currNode.data = string;
  }

  this.search = function(string) {
    let currNode = this.head;

    for (const char of string) {
      if (char in currNode.children) {
        currNode = currNode.children[char];
      } else {
        return false;
      }
    }

    if (currNode.data === string) {
      return true;
    } else {
      return false;
    }
  }
}

const trie = new Trie();
const stringArr = ['apple', 'appolo', 'apptite', 'appstore', 'appdulla'];

stringArr.forEach(string => {
  trie.insert(string);
})
  
console.log('apple: ', trie.search('apple'));
console.log('apple: ', trie.search('appolo'));
console.log('appgujeong: ', trie.search('apgujeong'));
